src/toast/ops/detrend.py

```
@@ -0,0 +1,172 @@
+        "linear",
+        help="Method to detrend. Valid methods are 'linear', 'mean', 'median'",
+    )
+
+    match_nsample = Int(
+        1,
+        help="When using 'linear' method, number of samples to match at the edge",
+    )
+
+    @traitlets.validate("shared_flag_mask")
+    def _check_shared_flag_mask(self, proposal):
+        check = proposal["value"]
+        if check < 0:
+            raise traitlets.TraitError("Shared flag mask should be a positive integer")
+        return check
+
+    @traitlets.validate("det_flag_mask")
+    def _check_det_flag_mask(self, proposal):
+        check = proposal["value"]
+        if check < 0:
+            raise traitlets.TraitError("Det flag mask should be a positive integer")
+        return check
+
+    @traitlets.validate("method")
+    def _check_method(self, proposal):
+        check = proposal["value"]
+        allowed = ["mean", "linear", "median"]
+        if check not in allowed:
+            msg = f"method must be one of {allowed}, not {check}"
+            raise traitlets.TraitError(msg)
+        return check
+
+    @traitlets.validate("match_nsample")
+    def _check_match_nsample(self, proposal):
+        check = proposal["value"]
+        if check < 1:
+            raise traitlets.TraitError("match_nsample should be a integer larger than 1")
+        return check
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+
+    @function_timer
+    def _exec(self, data, detectors=None, **kwargs):
+        log = Logger.get()
+
+        for ob in data.obs:
+            local_dets = ob.select_local_detectors(selection=detectors, flagmask=self.det_mask)
+            shared_flags = ob.shared[self.shared_flags].data & self.shared_flag_mask
+            if len(local_dets) == 0:
+                # Nothing to do for this observation
+                continue
+            for det in local_dets:
+                sig = ob.detdata[self.det_data][det]
+                assert np.ndim(sig) == 1
```

@tskisner how do we feel about `assert` in TOAST? It is not typical.

---

Yes, please raise an exception instead.

---

I forgot to remove this assertion line. Is `ob.detdata[self.det_data][det][:]` always 1d array? Is there cases it become multi dim array?

---

The detector signal always has a per-sample shape of zero (scalar).  Other detdata quantities (pointing quaternions, stokes weights, etc) have an array of values per sample.

`ob.detdata[self.det_data]["foo"]` returns a reference to the row of the internal storage for this detector.  So you don't need the `[:]` here.

If you use numpy-style "fancy indexing" like `ob.detdata[self.det_data][["foo", "bar", "blat"], 2:10]` then it will return a **list** of references to the underlying buffers for those detectors.  This is different from numpy arrays, where fancy indexing of non-contiguous rows makes a copy into a new array.  This behavior is intentional, to make sure we do not accidentally get a copy of the detector data and then modify that copy rather than the underlying buffer.